src/core/annotation.js

```diff
@@ -1985,23 +2027,101 @@ class WidgetAnnotation extends Annotation {
       );
     }
 
-    const font = await WidgetAnnotation._getFontData(
+    let font = await WidgetAnnotation._getFontData(
       evaluator,
       task,
       this.data.defaultAppearanceData,
       this._fieldResources.mergedResources
     );
-    const [defaultAppearance, fontSize] = this._computeFontSize(
-      totalHeight - defaultPadding,
-      totalWidth - 2 * hPadding,
-      value,
-      font,
-      lineCount
-    );
+
+    let defaultAppearance, fontSize, lineHeight;
+    const encodedLines = [];
+    let encodingError = false;
+    for (const line of lines) {
+      const encodedString = font.encodeString(line);
+      if (encodedString.length > 1) {
+        encodingError = true;
+      }
+      encodedLines.push(encodedString.join(""));
+    }
+
+    if (encodingError && intent & RenderingIntentFlag.SAVE) {
+      // We don't have a way to render the field, so we just rely on the
+      // /NeedAppearances trick to let the different sofware render correctly
+      // this pdf.
+      return { needAppearances: true };
+    }
+
+    // We check that the font is able to encode the string.
+    if (encodingError && this._isOffscreenCanvasSupported) {
+      // If it can't then we fallback on fake unicode font (mapped on sans-serif
+      // for the rendering).
+      // It means that a printed form can be rendered differently (it depends on
+      // the sans-serif font) but at leat we've something to render.
+      // In an ideal world the associated font should handle correctly the
+      // possible chars but a user can add a smiley or whatever.
+      // We could try to embed a font but it means that we must have access
+      // to the raw font file.
+      const fontFamily = this.data.comb ? "monospace" : "sans-serif";
+      const fakeUnicodeFont = new FakeUnicodeFont(evaluator.xref, fontFamily);
+      const resources = fakeUnicodeFont.createFontResources(lines.join(""));
+      const newFont = resources.getRaw("Font");
+
+      if (this._fieldResources.mergedResources.has("Font")) {
+        const oldFont = this._fieldResources.mergedResources.get("Font");
+        for (const key of newFont.getKeys()) {
+          oldFont.set(key, newFont.getRaw(key));
+        }
+      } else {
+        this._fieldResources.mergedResources.set("Font", newFont);
+      }
```

Comment:
Would this perhaps be generally safer, and more future-proof, if we used the `Dict.merge` functionality instead?
```suggestion
      this._fieldResources.mergedResources = Dict.merge({
        xref: evaluator.xref,
        dictArray: [this._fieldResources.mergedResources, resources],
        mergeSubDicts: true,
      });
```

Reply:
My first version was using it but it wasn't working correctly with `issue12233.pdf`.
If you enter `مرحبا بالعالم` in `Address 1` field and then print the pdf, you'll just get some garbage.
It's because the font is a ref and:
https://github.com/mozilla/pdf.js/blob/master/src/core/primitives.js#L267
Maybe we could fix that and use `merge` but I didn't want to add more complexity to this patch and spend too much time on something not directly related, so I think we could address that in an other patch.